examples/systemcalls/systemcalls.c

```diff
@@ -73,27 +143,75 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
 {
     va_list args;
     va_start(args, count);
+    pid_t process_id = -1;
+    pid_t wait_id = -1;
+    int32_t sys_status = -1;
+    int rc = -1;
     char * command[count+1];
     int i;
-    for(i=0; i<count; i++)
-    {
+    for(i=0; i<count; i++) {
         command[i] = va_arg(args, char *);
     }
     command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
-
-
 /*
  * TODO
  *   Call execv, but first using https://stackoverflow.com/a/13784315/1446624 as a refernce,
  *   redirect standard out to a file specified by outputfile.
  *   The rest of the behaviour is same as do_exec()
  *
 */
-
+    int fd = open(outputfile, O_WRONLY|O_TRUNC|O_CREAT, 0644);
+    if (fd < 0) {
+        perror("Assignment3:do_exec_redirect:open:");
+        return false; 
+    }
+    process_id = fork();
+    if (process_id == -1) {
+        perror("Assignment3:do_exec_redirect:fork:");
+        return false;
+    }
+    if (process_id == 0) {
+        
+        if (dup2(fd, 1) < 0) { 
+            perror("Assignment3:do_exec_redirect:dup2:");
+            exit(1); 
+        }
+        close(fd);
+        if(execv(command[0], &command[0]) == -1)
+        {
+                perror("Assignment3:do_exec_redirect:execv:");
+                exit(1);
+        }
+    }
+    else {
+        wait_id = waitpid(process_id, &sys_status, 0);
+        if (wait_id== -1) {
+            perror("Assignment3:do_exec:waitpid:");
+            rc = -1;
+            goto exit_prog;
+        }
+        if(wait_id != process_id) {
+            printf("Assignment3:do_exec:waitpid: ProcessID read is %d and exepected %d \n",
+                wait_id, process_id);
+            rc = -1;
+            goto exit_prog;
+        }
+        if (WIFEXITED (sys_status)) {
+            if(WEXITSTATUS (sys_status)) {
+                printf("Assignment3:do_exec:waitpid: Child process error code: %d \n",
+                    WEXITSTATUS(sys_status));
+                rc = -1;
+                goto exit_prog;
```

Comment:
This works to ensure cleanup paths.
See discussion in class about avoiding duplication between these two very similar functions.

Reply:
Thanks Dan, yes, I agree that duplication can be avoided. I was thinking about the same while implementing the solution. But was not sure if you wanted that level of optimization.